beacon-chain/rpc/eth/helpers/sync.go

```diff
@@ -53,26 +56,124 @@ func ValidateSync(
+			slotNumber, parseErr := strconv.ParseUint(stateIdString, 10, 64)
+			if parseErr != nil {
+				// ID format does not match any valid options.
+				e := statefetcher.NewStateIdParseError(parseErr)
+				return true, &e
+			}
+			optimistic, err := optimisticModeFetcher.IsOptimistic(ctx)
+			if err != nil {
+				return true, errors.Wrap(err, "could not check optimistic status")
+			}
+			if !optimistic {
+				return false, nil
+			}
+			finalizedSlot, err := slots.EpochStart(chainInfo.FinalizedCheckpt().Epoch)
+			if err != nil {
+				return true, errors.Wrap(err, "could not get head state's finalized slot")
+			}
+			if primitives.Slot(slotNumber) <= finalizedSlot {
+				return false, nil
+			}
+			if primitives.Slot(slotNumber) == chainInfo.HeadSlot() {
+				// We know the head is optimistic because we checked it above.
+				return true, nil
+			}
+			headRoot, err := chainInfo.HeadRoot(ctx)
+			if err != nil {
+				return true, errors.Wrap(err, "could not get head root")
+			}
+			r, err := chainInfo.ForkChoicer().AncestorRoot(ctx, bytesutil.ToBytes32(headRoot), primitives.Slot(slotNumber))
+			if err != nil {
+				return true, errors.Wrap(err, "could not get ancestor root")
+			}
+			return optimisticModeFetcher.IsOptimisticForRoot(ctx, r)
 		}
-		header.StateRoot = root[:]
 	}
-	headRoot, err := header.HashTreeRoot()
+}
+
+func isStateRootOptimistic(
+	ctx context.Context,
+	stateId []byte,
+	optimisticModeFetcher blockchain.OptimisticModeFetcher,
+	stateFetcher statefetcher.Fetcher,
+	chainInfo blockchain.ChainInfoFetcher,
+	database db.ReadOnlyDatabase,
+) (bool, error) {
+	st, err := stateFetcher.State(ctx, stateId)
 	if err != nil {
-		return false, errors.Wrap(err, "could not get header root")
+		return true, errors.Wrap(err, "could not fetch state")
 	}
-	isOptimistic, err := optimisticModeFetcher.IsOptimisticForRoot(ctx, headRoot)
+	if st.Slot() == chainInfo.HeadSlot() {
+		return optimisticModeFetcher.IsOptimistic(ctx)
+	}
+	_, roots, err := database.BlockRootsBySlot(ctx, st.Slot())
 	if err != nil {
-		return false, errors.Wrap(err, "could not check if block is optimistic")
+		return true, errors.Wrapf(err, "could not get block roots for slot %d", st.Slot())
+	}
+	if len(roots) == 0 {
+		return false, errors.New("no blocks returned from the database")
+	}
+	for _, r := range roots {
+		canonical, err := chainInfo.IsCanonical(ctx, r)
+		if err != nil {
+			return true, errors.Wrapf(err, "could not check canonical status")
+		}
+		if canonical {
+			optimistic, err := optimisticModeFetcher.IsOptimistic(ctx)
+			if err != nil {
+				return true, errors.Wrap(err, "could not check optimistic status")
+			}
+			if !optimistic {
+				return false, nil
+			}
+			finalizedSlot, err := slots.EpochStart(chainInfo.FinalizedCheckpt().Epoch)
+			if err != nil {
+				return true, errors.Wrap(err, "could not get head state's finalized slot")
+			}
+			if st.Slot() <= finalizedSlot {
+				return false, nil
+			}
+			return true, nil
+		}
```

Comment:
This block does not check what you want. Suppose you get `roots = [r1, r2]` with r1 canonical and r2 not, both past the finalized version. The node is not optimistic. In this situation you would return `false`. 

Now suppose that the block with root `r2` has `stateRoot`  precisely the state you queried for. In this case you should have returned `true`. 

What you need to do here is to requests the blocks, not the roots. And then check if the stateRoot matches the given id, and in that case (only in that case) check if it is canonical or not, etc...